Added support for the CLI's "--category" and "--entity-type" parameters.

MattLebrao opened the pull request:
Overall, only small changes were made to the project.

- Heartbeat.cs file:
    - `Heartbeat` class:
        - Added `Category` and `EntityType` properties.
    - `WakaTime.Shared.ExtensionUtils` namespace:
        - Added two enum types (`HearbeatCategory` and `EntityType`) for the valid values of CLI's "category" and "entity-type" parameters. 
        - See valid values in CLI file **wakatime/wakatime-cli/cmd/root.go** lines [45](https://github.com/wakatime/wakatime-cli/blob/435b6c667bc2e426f284cd1de38fb168ec82ebcf/cmd/root.go#L45) and [67](https://github.com/wakatime/wakatime-cli/blob/435b6c667bc2e426f284cd1de38fb168ec82ebcf/cmd/root.go#L67)


- WakaTime.cs file:
    - `WakaTime` class:
        - `HandleActivity` method:
            - Added two optional parameters (`activity` and `entityType`) for the `category` and `entityType` of the `AppendHeartbeat` method (optional to maintain backwards compatbility -- could be replaced with a method overload).
        - `AppendHeartbeat` method:
            - Added two parameters (`category` and `entityType`) for the the `Category` and `EntityType` properties of `Heartbeat` object (no need worry about backwards compatbility since this is a private member).

- PythonCliParameters.cs file:
    - `PythonCliParameters` class:
        - Added `Category` and `EntityType` properties.
        - `ToArray` method:
            - Included newly created properties' values with the proper CLI parameter names in the returning `parameters` collection.

- AssemblyInfo.cs file:
    - Set versions' revision numbers to +1.

MattLebrao commented on src/WakaTime.Shared.ExtensionUtils/Properties/AssemblyInfo.cs:
```diff
@@ -32,5 +32,5 @@
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyVersion("2.1.0")]
-[assembly: AssemblyFileVersion("2.1.0")]
+[assembly: AssemblyVersion("2.1.0.1")]
```
I've set the assembly version revision number as a mere suggestion, however, in retrospect, I'm not sure how NuGet managers determine when to auto update a library. Perhaps it will not automatically update if the version number is different, in which case changing the version number, even if just the revision number, could have been a mistake. (Specially considering it should still be binary-compatible with the current version).

alanhamlett replied:
We don't bump versions in PRs, let's leave this unchanged.

gandarez commented:
@MattLebrao I'll start the review ASAP. Have you joined us at Slack? After you sign up ping me there `Carlos Gandarez`.

https://wakatime.slack.com/

alanhamlett commented:
Use https://wakaslack.herokuapp.com/ to join Slack.

MattLebrao commented:
Version number was set back to the original. See 5afc95ae3d812c23f81679e056787086b76a18ca.

MattLebrao committed:
Added support for the "--category" and "-entity-type" CLI parameters.

Also added a way to make use of them through the "HandleAvtivity" method using optional method args to keep backwards compatbility.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Heartbeat.cs:
```diff
@@ -6,5 +6,43 @@ public class Heartbeat
+        public EntityType EntityType { get; set; }
+    }
+
+    /* Matheus Rocha (MattLebrao) on Aug 31 2020 @ 19h28:
```
I do prefer to not have any comment in the code since it's well documented in the client's repository.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Heartbeat.cs:
```diff
@@ -6,5 +6,43 @@ public class Heartbeat
+     *                                                 "manual testing", "code reviewing", "browsing" and "designing".
+     * Considering it has a set of valid values it makes sense to make an enum out of it, and use it with "ToString().Replace('_', ' ')".
+     */
+    public enum HeartbeatCategory
```
Let's move it to a dedicated class file called `Enums.cs` and then reference it here.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Heartbeat.cs:
```diff
@@ -6,5 +6,43 @@ public class Heartbeat
+     * Possible values for the entity type paramater are: "file", "domain", "app".
+     * Considering it has a set of valid values it makes sense to make an enum out of it, and use it with "ToString()".
+     */
+    public enum EntityType
```
Let's move it to a dedicated class file called `Enums.cs` and then reference it here.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Heartbeat.cs:
```diff
@@ -6,5 +6,43 @@ public class Heartbeat
+     */
+    public enum HeartbeatCategory
+    {
+        coding,
```
Enums use Pascal Case, So it should be `Coding`, `Building`.

MattLebrao replied:
Actually, I always write enums with Upper CamelCase (aka Pascal case).
I didn't do it here to match the parameters names.
I hadn't really considered using the description attribute then. Good idea

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Heartbeat.cs:
```diff
@@ -6,5 +6,43 @@ public class Heartbeat
+        building,
+        indexing,
+        debugging,
+        running_tests,
```
Let's remove the underscore and write it as `RunningTests` and then decorate with `Description`. Like:

```C#
public enum HeartbeatCategory
{
    [Description("running tests")]
    RunningTests
}
```

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Heartbeat.cs:
```diff
@@ -6,5 +6,43 @@ public class Heartbeat
+        designing
+    }
+
+    /* Matheus Rocha (MattLebrao) on Aug 31 2020 @ 21h32:
```
The same here.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Heartbeat.cs:
```diff
@@ -6,5 +6,43 @@ public class Heartbeat
+     */
+    public enum EntityType
+    {
+        file,
```
Let's use Pascal Case and `Description` attribute.

MattLebrao replied:
On the way

gandarez commented on src/WakaTime.Shared.ExtensionUtils/PythonCliParameters.cs:
```diff
@@ -31,7 +33,11 @@ public string[] ToArray(bool obfuscate = false)
-                Plugin
+                Plugin,
+                "--category",
+                Category.ToString().Replace('_', ' '),
```
Get the description here instead of replacing the character.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/PythonCliParameters.cs:
```diff
@@ -31,7 +33,11 @@ public string[] ToArray(bool obfuscate = false)
+                "--category",
+                Category.ToString().Replace('_', ' '),
+                "--entity-type",
+                EntityType.ToString()
```
The same here.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/WakaTime.cs:
```diff
@@ -78,7 +78,9 @@ public void InitializeAsync()
         }
 
-        public void HandleActivity(string currentFile, bool isWrite, string project)
+        //Added "activity" and "entityType" as optional parameters to preserve backwards compatbility.
```
Remove comment.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/WakaTime.cs:
```diff
@@ -91,17 +93,19 @@ public void HandleActivity(string currentFile, bool isWrite, string project)
+                Category = category,
+                EntityType = entityType
             };
             HeartbeatQueue.Enqueue(h);
```
Insert a break line here please.

MattLebrao replied:
You mean before or after `HeartbeatQueue.Enqueue(h);`?

gandarez reviewed (changes requested):
Left few comments.

MattLebrao committed:
Made changes requested in review.

- Enums were moved to a separate file.

- Enums are in Pascal case with description attributes.

- Passes enums' description to the CLI as parameter values.

- Removed comments from code.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Enums.cs:
```diff
@@ -0,0 +1,54 @@
+    {
+        [Description("coding")]
+        Coding,
+
```
Blank spaces here should be removed.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Enums.cs:
```diff
@@ -0,0 +1,54 @@
+        [Description("debugging")]
+        Debugging,
+
+        [Description("running_tests")]
```
The cli expects "running tests" (considering a space between word).

MattLebrao replied:
That is true, I forgot that. I wrote the description as the enum was named before.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Enums.cs:
```diff
@@ -0,0 +1,54 @@
+        [Description("running_tests")]
+        RunningTests,
+
+        [Description("writing_tests")]
```
The same here

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Enums.cs:
```diff
@@ -0,0 +1,54 @@
+        [Description("writing_tests")]
+        WritingTests,
+
+        [Description("manual_testing")]
```
The same here

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Enums.cs:
```diff
@@ -0,0 +1,54 @@
+        [Description("manual_testing")]
+        ManualTesting,
+
+        [Description("code_reviewing")]
```
The same here

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Enums.cs:
```diff
@@ -0,0 +1,54 @@
+    {
+        [Description("file")]
+        File,
+
```
Blank spaces should be removed.

gandarez commented on src/WakaTime.Shared.ExtensionUtils/Extensions.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace WakaTime.Shared.ExtensionUtils
+{
+    public static class Extensions
```
Let's rename it to `EnumExtensions` and also rename the matching file.

gandarez reviewed (changes requested):
Few other changes and will be ready to be pushed.

MattLebrao committed:
Made changes requested in review.

Included .csproj as it now lists two new files.

gandarez committed:
refactoring

gandarez commented:
@MattLebrao one last thing. Use this commit to apply the changes into yours. I already explained what to do in the Slack.

https://github.com/wakatime/nuget-wakatime-shared-extension-utils/commit/8738131e6c0aeb064d7e14da7cc0969d9e4b2b90

gandarez reviewed (approved):
Looks good
